model-optimizer/extensions/front/tf/KerasRNNReplacer.py

```diff
@@ -0,0 +1,237 @@
+            const_true = Const(body_graph, {'value': True}).create_node()
+            exec_cond_node.in_port(0).get_connection().set_source(const_true.out_port(0))
+
+
+class KerasLSTMReplacer(FrontReplacementSubgraph):
+    """
+    TensorFlow 2 Keras LSTM operation is expressed through a sub-graph with While operation with a certain body graph.
+    This replacer allows to detect such a sub-graph and to handle it by avoiding unsupported operations:
+    TensorListFromTensor, TensorListReserve, and TensorListStack in the main graph;
+    TensorListGetItem and TensorListSetItem in the body graph.
+    """
+    enabled = True
+
+    def run_before(self):
+        return [WhileNormalize]
+
+    @staticmethod
+    def pattern(**kwargs):
+        return dict(
+            nodes=[('unstack', dict(op='TensorListFromTensor')),
+                   ('reserve', dict(op='TensorListReserve')),
+                   ('while', dict(op='Loop')),
+                   ('stack', dict(op='TensorListStack')),
+                   ],
+            edges=[('reserve', 'while', {'in': 3}),
+                   ('unstack', 'while', {'in': 7}),
+                   ('while', 'stack', {'out': 3})]
+        )
+
+    def replace_sub_graph(self, graph: Graph, ext_match: dict):
+        int_match = find_intergraph_match(graph, ext_match, 3, 7, 3)
+        if int_match is not None:
+            process_keras_rnn(graph, ext_match, int_match, 6)
+
+
+class KerasGRUReplacer(FrontReplacementSubgraph):
+    """
+    TensorFlow 2 Keras GRU operation is expressed through a sub-graph with While operation with a certain body graph.
+    This replacer allows to detect such a sub-graph and to handle it by avoiding unsupported operations:
+    TensorListFromTensor, TensorListReserve, and TensorListStack in the main graph;
+    TensorListGetItem and TensorListSetItem in the body graph.
+    It transforms the sub-graph so that no list of tensors used, axis for Loop input ports performs slicing
+    (TensorListFromTensor->TensorListGetItem), axis for Loop output port for concatenation of outputs generated on each
+    iteration (TensorListSetItem->TensorListStack).
+    """
+    enabled = True
+
+    def run_before(self):
+        return [WhileNormalize]
+
+    @staticmethod
+    def pattern(**kwargs):
+        return dict(
+            nodes=[('unstack', dict(op='TensorListFromTensor')),
+                   ('reserve', dict(op='TensorListReserve')),
+                   ('while', dict(op='Loop')),
+                   ('stack', dict(op='TensorListStack')),
+                   ],
+            edges=[('reserve', 'while', {'in': 3}),
+                   ('unstack', 'while', {'in': 6}),
+                   ('while', 'stack', {'out': 3})]
+        )
+
+    def replace_sub_graph(self, graph: Graph, ext_match: dict):
+        int_match = find_intergraph_match(graph, ext_match, 3, 6, 3)
+        if int_match is not None:
+            process_keras_rnn(graph, ext_match, int_match, 5)
```

Comment:
The classes are the same except the port numbers. Let's implement a generic front transformation which will generate pattern for external pattern matching based on 3 values: 3, 6, 3 or 3, 7, 3. Take a look at the extensions/analysis/tf_yolo.py. But you case you will generate the dictionary returned from the `pattern` function based on parameters you specify (port numbers).

Reply:
Done. Now it avoids use of explicit values of port indices.